src/shared/models/k8sResource.ts

```diff
@@ -1,36 +1,24 @@
-import {Scalar} from 'yaml';
+import * as Rt from 'runtypes';
 
-import {KubernetesObject} from './appState';
-import {SarifRule} from './policy';
+import {K8sObject, K8sObjectRuntype} from './k8s';
+import {
+  AnyOrigin,
+  AnyOriginRuntype,
+  ClusterOrigin,
+  ClusterOriginRuntype,
+  HelmOrigin,
+  HelmOriginRuntype,
+  KustomizeOrigin,
+  KustomizeOriginRuntype,
+  LocalOrigin,
+  LocalOriginRuntype,
+} from './origin';
 
-type RefNode = {scalar: Scalar; key: string; parentKeyPath: string};
-
-type ResourceValidationError = {
-  property: string;
-  message: string;
-  errorPos?: RefPosition;
-  description?: string;
-  rule?: SarifRule;
-};
-
-type ResourceValidation = {
-  isValid: boolean;
-  errors: ResourceValidationError[];
-};
-
-/**
- * A k8s resource manifest, either extracted from a file or generated internally (for example when previewing kustomizations or helm charts)
- */
-type K8sResource = {
+export type ResourceMeta<Origin extends AnyOrigin = AnyOrigin> = {
```

Comment:
I think it's easier to create a type union which you can discriminate based on a property instead of working with generics. You can look at compare ResourceSet as I use the pattern there as well.

```
type Origin = LocalOrigin | ClusterOrigin | ..;

type BaseOrigin = {
  type: string;
  // other shared properties
};

type LocalOrigin = BaseOrigin & {
  type: "local",
  // specific properties
  directory: string;
}

// Now you can easily discriminate
if (resource.origin.type === "local") {
  resource.origin.directory // correctly infers.
}
```

I guess you use it here as you want to enforce that it always has the same origin? Asserting it from AnyOrigin to SpecificOrigin is a bit more tedious though so wanted to give an alternative approach.

Reply:
The initial version of this wasn't using generics but I was thinking there will be too many different types to remember and then when I switched to the generic it seemed much easier to read